File: 2015/js/day01.js
import { runDay } from "../../utils.js";

// https://adventofcode.com/2015/day/1

runDay(2015, 1)
  //
  .part(1, part1)
  .part(2, part2)
  .end();

function part1(inp) {
  let res = 0;
  for (let i = 0; i < inp.length; i++) {
    if (inp[i] === "(") {
      res++;
    } else {
      res--;
    }
  }
  return res;
}

function part2(inp) {
  let current = 0;
  for (let i = 0; i < inp.length; i++) {
    if (inp[i] === "(") {
      current++;
    } else {
      current--;
    }

    if (current === -1) return i + 1;
  }
}

File: 2016/python/day01.py
from aoc import run_day


def part1(inp):
    depths = inp.split(", ")

    directions = [
        [-1, 0],
        [0, 1],
        [1, 0],
        [0, -1],
    ]
    dir = 0
    current = (0, 0)

    for r in depths:
        if r[0] == "R":
            dir += 1
        else:
            dir -= 1

        current = (
            current[0] + directions[dir % 4][0] * int(r[1:]),
            current[1] + directions[dir % 4][1] * int(r[1:]),
        )

    return abs(current[0]) + abs(current[1])


def part2(inp):
    depths = inp.split(", ")

    directions = [
        [-1, 0],
        [0, 1],
        [1, 0],
        [0, -1],
    ]
    dir = 0
    current = (0, 0)
    visited = {}

    for r in depths:
        if r[0] == "R":
            dir += 1
        else:
            dir -= 1

        for i in range(int(r[1:])):
            current = (
                current[0] + directions[dir % 4][0],
                current[1] + directions[dir % 4][1],
            )

            if current in visited:
                return abs(current[0]) + abs(current[1])
            else:
                visited[current] = True


run_day(2016, 1).parts(
    [
        [1, part1],
        [2, part2],
    ]
).end()
